dio/lib/src/options.dart

```diff
@@ -587,10 +587,10 @@ class RequestOptions extends _RequestConfig with OptionsMixin {
   Uri get uri {
     String url = path;
     if (!url.startsWith(RegExp(r'https?:'))) {
-      url = baseUrl + url;
+      url = '$baseUrl/$url';
```

Comment:
Can you try to build the URL with the Uri class? I assume that should probably result in a safer implementation. It's constructor takes a list of path segments, so I guess it should also be easy to use.

Reply:
@ueman Sure! Since this is my first contribution to this library, I didn't want to be too intrusive, but your idea is really great! Therefore, to make the most of the Uri object, I would completely refactor the algorithm in the following way:

```dart
Uri get uri {
    Uri uri = Uri.parse(path);

    if (!uri.isAbsolute) {
      uri = Uri.parse(baseUrl).resolveUri(uri);
    }

    final query = Transformer.urlEncodeQueryMap(queryParameters, listFormat);
    if (query.isNotEmpty) {
      final querySeparator = uri.query.isNotEmpty ? '&' : '';
      final replacedQuery = '${uri.query}$querySeparator$query';
      uri = uri.replace(query: replacedQuery);
    }

    // Normalize the url.
    return uri.normalizePath();
  }
```

(Take this algorithm with a grain of salt: I haven't tested it yet 😅)

Reply:
That approach looks indeed much better. Thank you

Reply:
After applying the refactoring, I noticed a step in the old algorithm that may require further consideration. The following code snippet exists in the previous approach:

```dart
    if (query.isNotEmpty) {
      url += (url.contains('?') ? '&' : '?') + query;
    }
```

This logic checks if the URL already contains parameters, but it may not make sense to check for parameter presence in the URL since it should always be a path, and parameter handling is delegated to the `queryParameters` field. In other words, is it expected to have parameters included in the `baseUrl` or the `path`, or is it simply a defensive approach?

Reintroducing this logic raises a couple of concerns:

1. What should we do if a parameter in the path is encoded differently from `listFormat`?
2. What should we do if a parameter in the path also exists in `queryParameters`? Should we attempt to merge the two values or replace the value in `queryParameters`?

It would be prudent to carefully evaluate these issues and make an informed decision before reintroducing this logic in the refactoring.

@ueman @cfug-dev

Reply:
Are there existing tests for that part of the logic?

Reply:
@ueman no, unfortunately not...

Reply:
@kuhnroyal @AlexV525 What do you think?

Reply:
I haven't read this completely, but we should be cautious when writing new implementations that abandon old behavior. I would suggest that you take steps to change (in many PRs maybe):
1. Resolves duplicate `/` only, without additional changes.
2. Before the next major release, extract how we handle `ListFormat` as configurable.
3. In the next major release, apply the same handling to the `baseUrl`, and convert the implementation to `Uri`.